cluster/router/chain/chain.go

```diff
@@ -79,6 +103,99 @@ func (c *RouterChain) AddRouters(routers []router.PriorityRouter) {
 	c.routers = newRouters
 }
 
+// SetInvokers receives updated invokers from registry center. If the times of notification exceeds countThreshold and
+// time interval exceeds timeThreshold since last cache update, then notify to update the cache.
+func (c *RouterChain) SetInvokers(invokers []protocol.Invoker) {
+	c.invokers = invokers
+
+	c.count++
+	now := time.Now()
+	if c.count >= countThreshold && now.Sub(c.last) >= timeThreshold {
+		c.last = now
+		c.count = 0
+		go func() {
+			c.ch <- struct{}{}
+		}()
+	}
+}
+
+// loop listens on events to update the address cache when it's necessary, either when it receives notification
+// from address update, or when timeInterval exceeds.
+func (c *RouterChain) loop() {
+	for {
+		select {
+		case <-time.Tick(timeInterval):
+			c.buildCache()
+		case <-c.ch:
+			c.buildCache()
+		}
+	}
+}
+
+// copyRouters make a snapshot copy from RouterChain's router list.
+func (c *RouterChain) copyRouters() []router.PriorityRouter {
+	c.mutex.RLock()
+	ret := copySlice(c.routers)
+	c.mutex.RUnlock()
+	return ret.([]router.PriorityRouter)
+}
+
+// copyInvokers copies a snapshot of the received invokers.
+func (c *RouterChain) copyInvokers() []protocol.Invoker {
+	c.mutex.RLock()
+	ret := copySlice(c.invokers)
+	c.mutex.RUnlock()
+	return ret.([]protocol.Invoker)
+}
+
+func copySlice(s interface{}) interface{} {
+	t, v := reflect.TypeOf(s), reflect.ValueOf(s)
+	c := reflect.MakeSlice(t, v.Len(), v.Len())
+	reflect.Copy(c, v)
+	return c.Interface()
+}
+
+// loadCache loads cache from sync.Value to guarantee the visibility
+func (c *RouterChain) loadCache() *InvokerCache {
+	v := c.cache.Load()
+	if v == nil {
+		return nil
+	}
+
+	return v.(*InvokerCache)
+}
+
+// buildCache builds address cache with the new invokers for all poolable routers.
+func (c *RouterChain) buildCache() {
+	if c.invokers == nil || len(c.invokers) == 0 {
+		return
+	}
+
+	invokers := c.copyInvokers()
+	cache := BuildCache(invokers)
+	origin := c.loadCache()
+
+	var mutex sync.Mutex
+	var wg sync.WaitGroup
+
+	for _, r := range c.copyRouters() {
+		if p, ok := r.(router.Poolable); ok {
+			wg.Add(1)
+			go func(p router.Poolable) {
+				pool, info := poolRouter(p, origin, invokers)
+				mutex.Lock()
```

Comment:
You'd better use `defer mutex.Unlock()` to unlock safely.

Reply:
done.